File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.py
def solution(sequence):
    #펄스 수열 서로 다른거 세팅해서
    #각각 수열을 구해서 가장 합이 큰 부분을 구하자
    first = [(-1)**i for i in range(len(sequence))]
    second = [(-1)**(i+1) for i in range(len(sequence))]
    for i in range(len(sequence)):
        first[i] = sequence[i]*first[i]
        second[i] = sequence[i]*second[i]
        
    dp1 = [first[0]]
    dp2 = [second[0]]
    for i in range(1,len(sequence)):
        dp1.append(max(first[i],first[i]+dp1[i-1]))
        dp2.append(max(second[i],second[i]+dp2[i-1]))
    
    return max(max(dp1),max(dp2))

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
import java.util.*;
class Solution {
    public long solution(int[] sequence) {
        long[][] dp = new long[sequence.length][2];
        dp[0][0] = sequence[0];
        dp[0][1] = -sequence[0];
        
        for(int i=1;i<sequence.length;i++){
            dp[i][0] = Math.max(dp[i-1][1],0) + sequence[i];
            dp[i][1] = Math.max(dp[i-1][0],0) - sequence[i];
        }
        long ans = Long.MIN_VALUE;
        for(long[] c : dp){
            ans = Math.max(Math.max(c[0],c[1]),ans);
            // System.out.println(Arrays.toString(dp[i]));
        }
        
        return ans;
    }
}